packages/compartment-mapper/src/bundle.js

```diff
@@ -0,0 +1,333 @@
+    );
+  };
+
+  recur(entryCompartmentName, entryModuleSpecifier);
+
+  return modules;
+};
+
+/**
+ * @param {ReadFn} read
+ * @param {string} moduleLocation
+ * @param {Object} [options]
+ * @param {ModuleTransforms} [options.moduleTransforms]
+ * @returns {Promise<string>}
+ */
+export const makeBundle = async (read, moduleLocation, options) => {
+  const { moduleTransforms } = options || {};
+  const {
+    packageLocation,
+    packageDescriptorText,
+    packageDescriptorLocation,
+    moduleSpecifier,
+  } = await search(read, moduleLocation);
+
+  /** @type {Set<string>} */
+  const tags = new Set();
+
+  const packageDescriptor = parseLocatedJson(
+    packageDescriptorText,
+    packageDescriptorLocation,
+  );
+  const compartmentMap = await compartmentMapForNodeModules(
+    read,
+    packageLocation,
+    tags,
+    packageDescriptor,
+    moduleSpecifier,
+  );
+
+  const {
+    compartments,
+    entry: { compartment: entryCompartmentName, module: entryModuleSpecifier },
+  } = compartmentMap;
+  /** @type {Sources} */
+  const sources = {};
+
+  const makeImportHook = makeImportHookMaker(
+    read,
+    packageLocation,
+    sources,
+    compartments,
+  );
+
+  // Induce importHook to record all the necessary modules to import the given module specifier.
+  const { compartment, resolvers } = link(compartmentMap, {
+    resolve,
+    makeImportHook,
+    moduleTransforms,
+    parserForLanguage,
+  });
+  await compartment.load(entryModuleSpecifier);
+
+  const modules = sortedModules(
+    compartmentMap.compartments,
+    sources,
+    resolvers,
+    entryCompartmentName,
+    entryModuleSpecifier,
+  );
+
+  // Create an index of modules so we can resolve import specifiers to the
+  // index of the corresponding functor.
+  const modulesByKey = {};
+  for (let index = 0; index < modules.length; index += 1) {
+    const module = modules[index];
+    module.index = index;
+    modulesByKey[module.key] = module;
+  }
+  for (const module of modules) {
+    module.indexedImports = Object.fromEntries(
+      Object.entries(module.resolvedImports).map(([importSpecifier, key]) => [
+        importSpecifier,
+        modulesByKey[key].index,
+      ]),
+    );
+  }
+
+  // Only support mjs format.
+  const problems = modules
+    .filter(module => module.parser !== 'premjs')
+    .map(
+      ({ moduleSpecifier, compartmentName, parser }) =>
+        `module ${moduleSpecifier} in compartment ${compartmentName} in language ${parser}`,
+    );
+  if (problems.length) {
+    throw new Error(
+      `Can only bundle applications that only have ESM (.mjs-type) modules, got ${problems.join(
+        ', ',
+      )}`,
+    );
+  }
+
+  const bundle = `\
+(functors => {
+  function cell(name) {
+    const observers = [];
+    let value;
+    function set(newValue) {
+      value = newValue;
+      for (const observe of observers) {
+        observe(value);
```

Comment:
Not sure if it's worth it, but we may want to be defensive against misbehaving observers, at least wrapping in `try { } catch (e) { }`.